From 3ed97d1675f9390d29084f96a4580828b044cd5d Mon Sep 17 00:00:00 2001
From: Aaron Kling <webgeek1234@gmail.com>
Date: Wed, 24 Jan 2018 18:50:11 -0600
Subject: [PATCH 6/9] Allow building out of tree kernel modules

Based on an idea from the android-x86 mailing list

The kmod needs to define EXTRA_KERNEL_MODULE_PATH_$(LOCAL_MODULE)
as the source path. Then TARGET_EXTRA_KERNEL_MODULES is the list
of kmods to build.

Change-Id: I6175f3a4439a68497c2028c88aaf95beb2458b43
---
 build/tasks/kernel.mk | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/build/tasks/kernel.mk b/build/tasks/kernel.mk
index c4dcf6e..2ed291f 100644
--- a/build/tasks/kernel.mk
+++ b/build/tasks/kernel.mk
@@ -307,16 +307,25 @@ TARGET_KERNEL_BINARIES: $(KERNEL_CONFIG)
 	$(hide) if grep -q '^CONFIG_MODULES=y' $(KERNEL_CONFIG); then \
 			echo "Building Kernel Modules"; \
 			$(MAKE) $(MAKE_FLAGS) -C $(KERNEL_SRC) O=$(KERNEL_OUT) ARCH=$(KERNEL_ARCH) $(KERNEL_CROSS_COMPILE) $(KERNEL_CLANG_TRIPLE) $(KERNEL_CC) modules; \
+			$(foreach kmod,$(TARGET_EXTRA_KERNEL_MODULES),\
+				$(hide) rm -rf $(KERNEL_OUT)/extras/$(kmod); \
+				mkdir -p $(KERNEL_OUT)/extras/$(kmod); \
+				touch $(KERNEL_OUT)/extras/$(kmod)/Makefile; \
+				$(MAKE) $(MAKE_FLAGS) -C $(KERNEL_SRC) O=$(KERNEL_OUT) ARCH=$(KERNEL_ARCH) $(KERNEL_CROSS_COMPILE) $(KERNEL_CLANG_TRIPLE) $(KERNEL_CC) M=$(KERNEL_OUT)/extras/$(kmod) src=$(ANDROID_BUILD_TOP)/$(EXTRA_KERNEL_MODULE_PATH_$(kmod)) modules; \
+			) \
 		fi
 
 .PHONY: INSTALLED_KERNEL_MODULES
 INSTALLED_KERNEL_MODULES: depmod-host
 	$(hide) if grep -q '^CONFIG_MODULES=y' $(KERNEL_CONFIG); then \
 			echo "Installing Kernel Modules"; \
+			$(foreach kmod,$(TARGET_EXTRA_KERNEL_MODULES),\
+				$(MAKE) $(MAKE_FLAGS) -C $(KERNEL_SRC) O=$(KERNEL_OUT) ARCH=$(KERNEL_ARCH) $(KERNEL_CROSS_COMPILE) $(KERNEL_CLANG_TRIPLE) $(KERNEL_CC) INSTALL_MOD_PATH=../../$(KERNEL_MODULES_INSTALL) M=$(KERNEL_OUT)/extras/$(kmod) src=$(ANDROID_BUILD_TOP)/$(EXTRA_KERNEL_MODULE_PATH_$(kmod)) modules_install; \
+			) \
 			$(MAKE) $(MAKE_FLAGS) -C $(KERNEL_SRC) O=$(KERNEL_OUT) ARCH=$(KERNEL_ARCH) $(KERNEL_CROSS_COMPILE) $(KERNEL_CLANG_TRIPLE) $(KERNEL_CC) INSTALL_MOD_PATH=../../$(KERNEL_MODULES_INSTALL) modules_install && \
 			mofile=$$(find $(KERNEL_MODULES_OUT) -type f -name modules.order) && \
 			mpath=$$(dirname $$mofile) && \
-			for f in $$(find $$mpath/kernel -type f -name '*.ko'); do \
+			for f in $$(find $$mpath/kernel $$mpath/extra -type f -name '*.ko'); do \
 				$(KERNEL_TOOLCHAIN_PATH)strip --strip-unneeded $$f; \
 				mv $$f $(KERNEL_MODULES_OUT); \
 			done && \
-- 
2.7.4

